File: viddhana/packages/sdks/javascript/src/modules/ai.ts
/**
 * Prometheus AI Module - AI predictions and portfolio optimization
 */

import type {
  PricePrediction,
  PortfolioOptimization,
  RiskAssessment,
  RiskMetrics,
} from '../types';

/**
 * Prometheus AI module for predictions, optimization, and risk assessment
 *
 * @example
 * ```typescript
 * // Get price prediction
 * const prediction = await client.ai.predictPrice('BTC', 7);
 * console.log(`7-day trend: ${prediction.trend}`);
 *
 * // Optimize portfolio
 * const optimization = await client.ai.optimizePortfolio({
 *   userId: '0x...',
 *   portfolio: { BTC: 50000, ETH: 30000 },
 *   riskTolerance: 0.5,
 *   timeToGoal: 24,
 * });
 * ```
 */
export class PrometheusAI {
  private readonly apiUrl: string;
  private readonly apiKey?: string;

  constructor(apiUrl: string, apiKey?: string) {
    this.apiUrl = apiUrl;
    this.apiKey = apiKey;
  }

  /**
   * Makes an authenticated API request
   * @private
   */
  private async request<T>(endpoint: string, body: object): Promise<T> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };

    if (this.apiKey) {
      headers['Authorization'] = `Bearer ${this.apiKey}`;
    }

    const response = await fetch(`${this.apiUrl}${endpoint}`, {
      method: 'POST',
      headers,
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`API error ${response.status}: ${errorText}`);
    }

    return response.json() as Promise<T>;
  }

  /**
   * Gets price prediction for an asset
   *
   * @param asset - Asset symbol (e.g., 'BTC', 'ETH')
   * @param horizon - Number of days to predict (1-30)
   * @returns Price prediction with daily forecasts and confidence
   */
  async predictPrice(
    asset: string,
    horizon = 7
  ): Promise<PricePrediction> {
    return this.request<PricePrediction>('/v1/predict/price', {
      asset,
      horizon,
    });
  }

  /**
   * Gets portfolio optimization recommendation
   *
   * @param params - Optimization parameters
   * @returns Portfolio optimization with recommended actions
   */
  async optimizePortfolio(params: {
    userId: string;
    portfolio: Record<string, number>;
    riskTolerance: number; // 0-1
    timeToGoal: number; // months
  }): Promise<PortfolioOptimization> {
    return this.request<PortfolioOptimization>('/v1/optimize/portfolio', {
      user_id: params.userId,
      portfolio: params.portfolio,
      risk_tolerance: params.riskTolerance,
      time_to_goal: params.timeToGoal,
    });
  }

  /**
   * Assesses portfolio risk
   *
   * @param portfolio - Portfolio as { asset: value } mapping
   * @returns Risk assessment with metrics and recommendations
   */
  async assessRisk(
    portfolio: Record<string, number>
  ): Promise<RiskAssessment> {
    return this.request<RiskAssessment>('/v1/assess/risk', {
      portfolio,
    });
  }

  /**
   * Gets risk metrics for a portfolio
   *
   * @param portfolio - Portfolio as { asset: value } mapping
   * @returns Risk metrics including VaR, Sharpe ratio, etc.
   */
  async getRiskMetrics(
    portfolio: Record<string, number>
  ): Promise<RiskMetrics> {
    const assessment = await this.assessRisk(portfolio);
    return assessment.metrics;
  }

  /**
   * Gets market sentiment for an asset
   *
   * @param asset - Asset symbol
   * @returns Sentiment analysis
   */
  async getSentiment(asset: string): Promise<{
    asset: string;
    sentiment: 'bullish' | 'bearish' | 'neutral';
    score: number;
    sources: { name: string; sentiment: string; weight: number }[];
    timestamp: number;
  }> {
    return this.request('/v1/sentiment', { asset });
  }

  /**
   * Gets correlation analysis between assets
   *
   * @param assets - Array of asset symbols
   * @returns Correlation matrix
   */
  async getCorrelation(assets: string[]): Promise<{
    assets: string[];
    matrix: number[][];
    period: string;
    timestamp: number;
  }> {
    return this.request('/v1/correlation', { assets });
  }

  /**
   * Gets volatility forecast for an asset
   *
   * @param asset - Asset symbol
   * @param days - Number of days to forecast
   * @returns Volatility forecast
   */
  async getVolatilityForecast(
    asset: string,
    days = 7
  ): Promise<{
    asset: string;
    currentVolatility: number;
    forecast: { day: number; volatility: number; confidence: number }[];
    trend: 'increasing' | 'decreasing' | 'stable';
    timestamp: number;
  }> {
    return this.request('/v1/volatility/forecast', { asset, days });
  }

  /**
   * Streams real-time predictions via WebSocket
   *
   * @param asset - Asset symbol to stream predictions for
   * @param onUpdate - Callback for prediction updates
   * @param onError - Optional error callback
   * @returns Cleanup function to close the connection
   */
  streamPredictions(
    asset: string,
    onUpdate: (prediction: PricePrediction) => void,
    onError?: (error: Error) => void
  ): () => void {
    const wsUrl = this.apiUrl
      .replace('https://', 'wss://')
      .replace('http://', 'ws://');
    
    const ws = new WebSocket(`${wsUrl}/ws/predictions/${asset}`);

    ws.onopen = () => {
      // Send authentication if API key is present
      if (this.apiKey) {
        ws.send(JSON.stringify({ type: 'auth', apiKey: this.apiKey }));
      }
    };

    ws.onmessage = (event: MessageEvent) => {
      try {
        const prediction = JSON.parse(event.data as string) as PricePrediction;
        onUpdate(prediction);
      } catch (error) {
        if (onError) {
          onError(
            error instanceof Error ? error : new Error('Failed to parse prediction')
          );
        }
      }
    };

    ws.onerror = () => {
      if (onError) {
        onError(new Error('WebSocket connection error'));
      }
    };

    ws.onclose = () => {
      // Connection closed
    };

    // Return cleanup function
    return () => {
      ws.close();
    };
  }

  /**
   * Streams portfolio updates via WebSocket
   *
   * @param address - User address to monitor
   * @param onUpdate - Callback for portfolio updates
   * @param onError - Optional error callback
   * @returns Cleanup function to close the connection
   */
  streamPortfolio(
    address: string,
    onUpdate: (update: { type: string; data: unknown }) => void,
    onError?: (error: Error) => void
  ): () => void {
    const wsUrl = this.apiUrl
      .replace('https://', 'wss://')
      .replace('http://', 'ws://');
    
    const ws = new WebSocket(`${wsUrl}/ws/portfolio/${address}`);

    ws.onopen = () => {
      if (this.apiKey) {
        ws.send(JSON.stringify({ type: 'auth', apiKey: this.apiKey }));
      }
    };

    ws.onmessage = (event: MessageEvent) => {
      try {
        const update = JSON.parse(event.data as string) as {
          type: string;
          data: unknown;
        };
        onUpdate(update);
      } catch (error) {
        if (onError) {
          onError(
            error instanceof Error ? error : new Error('Failed to parse update')
          );
        }
      }
    };

    ws.onerror = () => {
      if (onError) {
        onError(new Error('WebSocket connection error'));
      }
    };

    return () => {
      ws.close();
    };
  }

  /**
   * Batch prediction for multiple assets
   *
   * @param assets - Array of asset symbols
   * @param horizon - Number of days to predict
   * @returns Array of predictions
   */
  async batchPredictPrice(
    assets: string[],
    horizon = 7
  ): Promise<PricePrediction[]> {
    const response = await this.request<{ predictions: PricePrediction[] }>(
      '/v1/predict/batch',
      { assets, horizon }
    );
    return response.predictions;
  }

  /**
   * Gets model information
   *
   * @returns AI model metadata
   */
  async getModelInfo(): Promise<{
    version: string;
    lastUpdated: number;
    accuracy: { price: number; volatility: number };
    supportedAssets: string[];
  }> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };

    if (this.apiKey) {
      headers['Authorization'] = `Bearer ${this.apiKey}`;
    }

    const response = await fetch(`${this.apiUrl}/v1/model/info`, {
      method: 'GET',
      headers,
    });

    if (!response.ok) {
      throw new Error(`API error: ${response.status}`);
    }

    return response.json() as Promise<{
      version: string;
      lastUpdated: number;
      accuracy: { price: number; volatility: number };
      supportedAssets: string[];
    }>;
  }
}

File: viddhana/packages/sdks/python/src/viddhana/ai.py
"""AI module for Prometheus AI predictions and optimizations."""

from typing import Dict, List, Optional
import httpx

from .types import (
    PricePrediction,
    PricePredictionPoint,
    PortfolioOptimization,
    RiskAssessment,
)


class AIModule:
    """Module for interacting with Prometheus AI services."""

    def __init__(self, api_url: str, api_key: Optional[str] = None):
        """
        Initialize AIModule.

        Args:
            api_url: Base URL for the AI API.
            api_key: Optional API key for authentication.
        """
        self._api_url = api_url.rstrip("/")
        self._api_key = api_key
        self._headers = self._build_headers()
        self._client = httpx.Client(
            base_url=self._api_url,
            headers=self._headers,
            timeout=30.0,
        )
        self._async_client: Optional[httpx.AsyncClient] = None

    def _build_headers(self) -> Dict[str, str]:
        """Build request headers."""
        headers = {"Content-Type": "application/json"}
        if self._api_key:
            headers["Authorization"] = f"Bearer {self._api_key}"
        return headers

    def _get_async_client(self) -> httpx.AsyncClient:
        """Get or create async HTTP client."""
        if self._async_client is None:
            self._async_client = httpx.AsyncClient(
                base_url=self._api_url,
                headers=self._headers,
                timeout=30.0,
            )
        return self._async_client

    def predict_price(self, asset: str, horizon: int = 7) -> PricePrediction:
        """
        Get price prediction for an asset.

        Args:
            asset: Asset symbol (e.g., 'BTC', 'ETH').
            horizon: Number of days to predict ahead (1-30).

        Returns:
            PricePrediction with daily predictions and confidence scores.

        Raises:
            httpx.HTTPStatusError: If the API request fails.
        """
        response = self._client.post(
            "/v1/predict/price",
            json={"asset": asset, "horizon": horizon},
        )
        response.raise_for_status()
        data = response.json()

        predictions = [
            PricePredictionPoint(
                day=p["day"],
                price=p["price"],
                confidence=p["confidence"],
            )
            for p in data.get("predictions", [])
        ]

        return PricePrediction(
            asset=data["asset"],
            current_price=data["currentPrice"],
            horizon=data["horizon"],
            predictions=predictions,
            trend=data["trend"],
            volatility_forecast=data["volatilityForecast"],
            model_version=data["modelVersion"],
            generated_at=data["generatedAt"],
        )

    async def predict_price_async(self, asset: str, horizon: int = 7) -> PricePrediction:
        """
        Async version of predict_price.

        Args:
            asset: Asset symbol (e.g., 'BTC', 'ETH').
            horizon: Number of days to predict ahead (1-30).

        Returns:
            PricePrediction with daily predictions and confidence scores.
        """
        client = self._get_async_client()
        response = await client.post(
            "/v1/predict/price",
            json={"asset": asset, "horizon": horizon},
        )
        response.raise_for_status()
        data = response.json()

        predictions = [
            PricePredictionPoint(
                day=p["day"],
                price=p["price"],
                confidence=p["confidence"],
            )
            for p in data.get("predictions", [])
        ]

        return PricePrediction(
            asset=data["asset"],
            current_price=data["currentPrice"],
            horizon=data["horizon"],
            predictions=predictions,
            trend=data["trend"],
            volatility_forecast=data["volatilityForecast"],
            model_version=data["modelVersion"],
            generated_at=data["generatedAt"],
        )

    def optimize_portfolio(
        self,
        user_id: str,
        portfolio: Dict[str, float],
        risk_tolerance: float,
        time_to_goal: int,
    ) -> PortfolioOptimization:
        """
        Get portfolio optimization recommendation.

        Args:
            user_id: User identifier (usually wallet address).
            portfolio: Current portfolio as {asset: value} dict.
            risk_tolerance: Risk tolerance 0-1 (0=conservative, 1=aggressive).
            time_to_goal: Months until investment goal.

        Returns:
            PortfolioOptimization with recommended actions.

        Raises:
            httpx.HTTPStatusError: If the API request fails.
        """
        response = self._client.post(
            "/v1/optimize/portfolio",
            json={
                "user_id": user_id,
                "portfolio": portfolio,
                "risk_tolerance": risk_tolerance,
                "time_to_goal": time_to_goal,
            },
        )
        response.raise_for_status()
        data = response.json()

        return PortfolioOptimization(
            action=data["action"],
            recommendations=data["recommendations"],
            confidence=data["confidence"],
            risk_assessment=data["risk_assessment"],
        )

    async def optimize_portfolio_async(
        self,
        user_id: str,
        portfolio: Dict[str, float],
        risk_tolerance: float,
        time_to_goal: int,
    ) -> PortfolioOptimization:
        """
        Async version of optimize_portfolio.

        Args:
            user_id: User identifier (usually wallet address).
            portfolio: Current portfolio as {asset: value} dict.
            risk_tolerance: Risk tolerance 0-1.
            time_to_goal: Months until investment goal.

        Returns:
            PortfolioOptimization with recommended actions.
        """
        client = self._get_async_client()
        response = await client.post(
            "/v1/optimize/portfolio",
            json={
                "user_id": user_id,
                "portfolio": portfolio,
                "risk_tolerance": risk_tolerance,
                "time_to_goal": time_to_goal,
            },
        )
        response.raise_for_status()
        data = response.json()

        return PortfolioOptimization(
            action=data["action"],
            recommendations=data["recommendations"],
            confidence=data["confidence"],
            risk_assessment=data["risk_assessment"],
        )

    def assess_risk(self, portfolio: Dict[str, float]) -> RiskAssessment:
        """
        Assess portfolio risk metrics.

        Args:
            portfolio: Portfolio as {asset: value} dict.

        Returns:
            RiskAssessment with risk metrics and recommendations.

        Raises:
            httpx.HTTPStatusError: If the API request fails.
        """
        response = self._client.post(
            "/v1/assess/risk",
            json={"portfolio": portfolio},
        )
        response.raise_for_status()
        data = response.json()

        return RiskAssessment(
            risk_score=data["risk_score"],
            volatility=data["volatility"],
            sharpe_ratio=data["sharpe_ratio"],
            max_drawdown=data["max_drawdown"],
            recommendations=data["recommendations"],
        )

    async def assess_risk_async(self, portfolio: Dict[str, float]) -> RiskAssessment:
        """
        Async version of assess_risk.

        Args:
            portfolio: Portfolio as {asset: value} dict.

        Returns:
            RiskAssessment with risk metrics and recommendations.
        """
        client = self._get_async_client()
        response = await client.post(
            "/v1/assess/risk",
            json={"portfolio": portfolio},
        )
        response.raise_for_status()
        data = response.json()

        return RiskAssessment(
            risk_score=data["risk_score"],
            volatility=data["volatility"],
            sharpe_ratio=data["sharpe_ratio"],
            max_drawdown=data["max_drawdown"],
            recommendations=data["recommendations"],
        )

    def close(self):
        """Close HTTP clients."""
        self._client.close()

    async def aclose(self):
        """Close async HTTP client."""
        if self._async_client:
            await self._async_client.aclose()
            self._async_client = None
